src/VisualStudio/Core/Def/ExternalAccess/LegacyCodeAnalysis/Internal/LegacyCodeAnalysisVisualStudioDiagnosticAnalyzerServiceAccessor.cs

```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Diagnostics;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.LegacyCodeAnalysis.Internal
```

Comment:
I think we want the implementation to be under a different namespace, not `Microsoft.CodeAnalysis.ExternalAccess.LegacyCodeAnalysis`. Otherwise, the internal implementation types from here will be visible to the partners, which is not desirable. How about `Microsoft.CodeAnalysis.ExternalAccess.Implementation.LegacyCodeAnalysis`?

Reply:
I'm planning to update the analyzer to ignore nested namespaces that are named `Internal`, `Implementation`, and/or `Private`

Reply:
I don't think that is the desired design, and does not sound intuitive at all. Adding @tmat @heejaechang @jinujoseph 

Basically your proposed design implies that users would have add a separate RestrictedIVT for every single child namespace that needs to be given external access. I think the design we decided is the following (which the analyzer already implements):

If user adds RestrictedIVT to Namespace "N1", that gives external access to internal types defined in N1 and _all the contained namepsaces_ and types. So if you had 3 types namespace N1.T1, N1.N2.T2, N1.N3.T3, you need a single restricted IVT with "N1" as the allowed namespace, not 3 separate Restricted IVTs, one each for "N1", "N1.N2" and "N1.N3". Latter will add lot of clutter and becomes unmanageable.

Reply:
I'm fine with an alternative proposal that meets the following conditions:

1. The namespace/type hierarchy matches the file layout on disk.
1. Each project has at most one folder for which 1) the content of the folder (nested) is owned by one partner and 2) the parent folder contains content not owned by that partner. (in this pull request, this folder is **src/VisualStudio/Core/Def/ExternalAccess/LegacyCodeAnalysis**)
1. Each project has at most one new top-level folder containing partner-owned code (in this pull request, this folder is **src/VisualStudio/Core/Def/ExternalAccess**).

Reply:
I would suggest the following folder hierarcy:
1. Top level folder for all partners: `src/VisualStudio/Core/Def/External`
2. Top level folder for specific partner: `src/VisualStudio/Core/Def/External/Partner`
3. Folder for restricted IVT APIs for specific partner: `src/VisualStudio/Core/Def/External/Partner/ExternalAccess` or `src/VisualStudio/Core/Def/External/Partner/ExternalApis` or any suitable name indicating supported external APIs only.

Equivalent namespaces would be:
1. `Microsoft.CodeAnalysis.External`
2. Specific partner namespace: `Microsoft.CodeAnalysis.External.Partner`
3. Restricted IVT namespace for partner: `Microsoft.CodeAnalysis.External.Partner.ExternalAccess` or `Microsoft.CodeAnalysis.External.Partner.ExternalApis`

Reply:
I like @mavasani's proposal better. I would rather not have any additional conventions assumed by the analyzer. Such would be hard to transfer to generally applicable compiler feature, if we were to move this validation to the compiler some day.